File: src/lib/types/ai.ts
/**
 * AI request/stream contract between the TS intelligence layer and the Rust
 * provider layer. Rust owns credentials + HTTP; TS owns prompts + policy.
 * Mirrors `bluey_core::types::ai`.
 */

import type { TraceStamps } from "./latency";
import type { BlueyError } from "./errors";
import type { ModelRole } from "./mode";
import type { Citation } from "./response";

export type AITask =
  | "vision"
  | "classification"
  | "transcription"
  | "answer"
  | "coding"
  | "system_design"
  | "summarization"
  | "research"
  | "deep_reasoning"
  | "embedding";

/**
 * The shape the answer should take — detected from the question and the
 * screen (`classifyIntent`), rendered as the `Shape:` line after `Task:` and
 * used for output budgets. Orthogonal to `AITask` (what kind of work) and to
 * the response schema (which fields come back). Never sent to Rust.
 */
export type AnswerShape =
  | "choice" // multiple choice: the option, one clause of why
  | "boolean" // yes/no, true/false
  | "fill_in" // fill in the blank
  | "calculation" // the result, then the working
  | "compare" // which of two is better and why
  | "short_answer" // one to three sentences
  | "explain" // answer first, then the reasons or steps
  | "spoken" // exactly what to say, first person
  | "written" // the text to send or submit
  | "code" // the working solution
  | "design" // a system design with its trade-offs
  | "summary"; // the points themselves

export type LatencyBudget = "ultra-fast" | "fast" | "balanced" | "deep";
export type ReasoningLevel = "none" | "light" | "deep";

/**
 * `google_gemini` is the default provider (ADR 0007); the others are alternates.
 * `chatgpt_codex`, `claude_subscription` and `antigravity_google` are served by
 * subscription accounts rather than API keys (ADR 0009; reserved ids `chatgpt`,
 * `claude`, `antigravity`).
 */
export type AIProviderKind =
  | "google_gemini"
  | "azure_foundry"
  | "anthropic"
  | "openai_compatible"
  | "chatgpt_codex"
  | "claude_subscription"
  | "antigravity_google"
  | "mock";

/** What texts are embedded for (documents get `title:` prefixes, queries `task:` prefixes on gemini-embedding-2). */
export type EmbedPurpose = "document" | "query";

export interface AIProviderConfig {
  id: string;
  kind: AIProviderKind;
  name: string;
  /** e.g. https://my-resource.openai.azure.com or https://api.anthropic.com */
  baseUrl: string;
  /** Azure api-version when using the legacy deployment endpoint. */
  apiVersion?: string;
  /** For Azure: map of logical model name -> deployment name (optional). */
  deployments?: Record<string, string>;
  enabled: boolean;
  /** True when a key is stored in the OS keychain (never the key itself). */
  hasApiKey: boolean;
  /** API key (default when absent) or an OAuth subscription account (ADR 0009). */
  authMethod?: "api_key" | "oauth_subscription";
}

export interface ModelAssignment {
  providerId: string;
  model: string;
}

export type ModelRoleAssignments = Record<ModelRole, ModelAssignment | null>;

export type AIContentPart =
  | { type: "text"; text: string }
  | { type: "image"; mediaType: "image/jpeg" | "image/png" | "image/webp"; data: string };

export interface AIMessage {
  role: "system" | "user" | "assistant";
  content: AIContentPart[];
}

export interface JsonSchemaSpec {
  name: string;
  schema: Record<string, unknown>;
  strict?: boolean;
}

export interface AIRequest {
  requestId: string;
  sessionId?: string;
  /** Monotonic per-window generation for stale-response protection. */
  generation: number;
  task: AITask;
  latencyBudget: LatencyBudget;
  reasoning: ReasoningLevel;
  visionRequired: boolean;
  /** Estimated input tokens (helps routing). */
  contextTokens: number;
  messages: AIMessage[];
  outputSchema?: JsonSchemaSpec;
  maxOutputTokens?: number;
  temperature?: number;
  /** Explicit override of the router decision. */
  modelOverride?: ModelAssignment;
  /** The WebView's half of the fast-path trace (ADR 0010 §2), when the engine measured it. */
  trace?: TraceStamps;
  createdAt: string;
}

export interface ModelSelection {
  providerId: string;
  providerKind: AIProviderKind;
  model: string;
  role: ModelRole;
  reason: string;
}

export type AIChunk =
  | { type: "started"; requestId: string; selection: ModelSelection }
  | { type: "delta"; requestId: string; text: string }
  | { type: "usage"; requestId: string; inputTokens?: number; outputTokens?: number }
  | {
      type: "completed";
      requestId: string;
      finishReason: "stop" | "length" | "cancelled" | "error";
      totalMs: number;
      timeToFirstTokenMs?: number;
    }
  | { type: "failed"; requestId: string; error: BlueyError };

export interface AIResponse {
  requestId: string;
  selection: ModelSelection;
  text: string;
  /** Parsed JSON when `outputSchema` was requested and parsing succeeded. */
  json?: unknown;
  finishReason: "stop" | "length" | "cancelled" | "error";
  inputTokens?: number;
  outputTokens?: number;
  timeToFirstTokenMs?: number;
  totalMs: number;
}

export interface ConnectionTestResult {
  ok: boolean;
  providerId: string;
  model?: string;
  latencyMs?: number;
  error?: BlueyError;
}

/** Research */
export type ResearchDepth = "none" | "search" | "search_scrape" | "deep_agent";

export interface SearchResult {
  id: string;
  title: string;
  url: string;
  snippet?: string;
  publishedAt?: string;
  source: "exa" | "firecrawl" | "mock";
}

export interface ScrapeResult {
  url: string;
  title?: string;
  markdown: string;
  source: "firecrawl" | "mock";
}

export interface DeepResearchRequest {
  jobId: string;
  sessionId?: string;
  /** PUBLIC query only — never include private context. */
  query: string;
  goal: string;
  maxTurns?: number;
  tools: Array<"exa_search" | "firecrawl_scrape" | "document_read">;
  /** Document ids the agent may read via the document_read tool (local, private). */
  allowedDocumentIds?: string[];
}

export type DeepResearchEvent =
  | { type: "started"; jobId: string }
  | { type: "progress"; jobId: string; message: string }
  | { type: "tool_call"; jobId: string; tool: string; input: Record<string, unknown> }
  | { type: "text_delta"; jobId: string; text: string }
  | {
      type: "completed";
      jobId: string;
      report: string;
      citations: Citation[];
      totalMs: number;
      turns: number;
    }
  | { type: "failed"; jobId: string; error: BlueyError };

File: src-tauri/crates/bluey-core/src/types/ai.rs
use serde::{Deserialize, Serialize};

use super::mode::ModelRole;
use super::response::Citation;
use crate::error::BlueyError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AiTask {
    Vision,
    Classification,
    Transcription,
    Answer,
    Coding,
    SystemDesign,
    Summarization,
    Research,
    DeepReasoning,
    Embedding,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "kebab-case")]
pub enum LatencyBudget {
    UltraFast,
    #[default]
    Fast,
    Balanced,
    Deep,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum ReasoningLevel {
    #[default]
    None,
    Light,
    Deep,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AiProviderKind {
    /// Google AI Studio (Gemini API) — the default provider (ADR 0007).
    GoogleGemini,
    AzureFoundry,
    Anthropic,
    OpenaiCompatible,
    Mock,
}

/// Mirrors `AIProviderConfig`. Never contains the API key (kept in the OS keychain).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiProviderConfig {
    pub id: String,
    pub kind: AiProviderKind,
    pub name: String,
    pub base_url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub api_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub deployments: Option<std::collections::BTreeMap<String, String>>,
    pub enabled: bool,
    pub has_api_key: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelAssignment {
    pub provider_id: String,
    pub model: String,
}

/// Mirrors `ModelRoleAssignments` (`Record<ModelRole, ModelAssignment | null>`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ModelRoleAssignments {
    pub default: Option<ModelAssignment>,
    pub fast: Option<ModelAssignment>,
    pub reasoning: Option<ModelAssignment>,
    pub vision: Option<ModelAssignment>,
    pub research: Option<ModelAssignment>,
    pub transcription: Option<ModelAssignment>,
    pub embedding: Option<ModelAssignment>,
}

impl ModelRoleAssignments {
    pub fn get(&self, role: ModelRole) -> Option<&ModelAssignment> {
        match role {
            ModelRole::Default => self.default.as_ref(),
            ModelRole::Fast => self.fast.as_ref(),
            ModelRole::Reasoning => self.reasoning.as_ref(),
            ModelRole::Vision => self.vision.as_ref(),
            ModelRole::Research => self.research.as_ref(),
            ModelRole::Transcription => self.transcription.as_ref(),
            ModelRole::Embedding => self.embedding.as_ref(),
        }
    }

    pub fn set(&mut self, role: ModelRole, assignment: Option<ModelAssignment>) {
        match role {
            ModelRole::Default => self.default = assignment,
            ModelRole::Fast => self.fast = assignment,
            ModelRole::Reasoning => self.reasoning = assignment,
            ModelRole::Vision => self.vision = assignment,
            ModelRole::Research => self.research = assignment,
            ModelRole::Transcription => self.transcription = assignment,
            ModelRole::Embedding => self.embedding = assignment,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ImageMediaType {
    #[serde(rename = "image/jpeg")]
    Jpeg,
    #[serde(rename = "image/png")]
    Png,
    #[serde(rename = "image/webp")]
    Webp,
}

impl ImageMediaType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Jpeg => "image/jpeg",
            Self::Png => "image/png",
            Self::Webp => "image/webp",
        }
    }
}

/// Mirrors `AIContentPart` (tag = "type").
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum AiContentPart {
    Text {
        text: String,
    },
    #[serde(rename_all = "camelCase")]
    Image {
        media_type: ImageMediaType,
        data: String,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AiRole {
    System,
    User,
    Assistant,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiMessage {
    pub role: AiRole,
    pub content: Vec<AiContentPart>,
}

impl AiMessage {
    pub fn text(role: AiRole, text: impl Into<String>) -> Self {
        Self {
            role,
            content: vec![AiContentPart::Text { text: text.into() }],
        }
    }

    pub fn has_images(&self) -> bool {
        self.content
            .iter()
            .any(|p| matches!(p, AiContentPart::Image { .. }))
    }

    pub fn text_content(&self) -> String {
        self.content
            .iter()
            .filter_map(|p| match p {
                AiContentPart::Text { text } => Some(text.as_str()),
                _ => None,
            })
            .collect::<Vec<_>>()
            .join("\n")
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct JsonSchemaSpec {
    pub name: String,
    pub schema: serde_json::Value,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub strict: Option<bool>,
}

/// Mirrors `AIRequest`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiRequest {
    pub request_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub generation: u64,
    pub task: AiTask,
    pub latency_budget: LatencyBudget,
    pub reasoning: ReasoningLevel,
    pub vision_required: bool,
    pub context_tokens: u32,
    pub messages: Vec<AiMessage>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_schema: Option<JsonSchemaSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_output_tokens: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model_override: Option<ModelAssignment>,
    pub created_at: String,
}

/// Mirrors `ModelSelection`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelSelection {
    pub provider_id: String,
    pub provider_kind: AiProviderKind,
    pub model: String,
    pub role: ModelRole,
    pub reason: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum FinishReason {
    Stop,
    Length,
    Cancelled,
    Error,
}

/// Mirrors `AIChunk` (tag = "type").
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AiChunk {
    #[serde(rename_all = "camelCase")]
    Started {
        request_id: String,
        selection: ModelSelection,
    },
    #[serde(rename_all = "camelCase")]
    Delta { request_id: String, text: String },
    #[serde(rename_all = "camelCase")]
    Usage {
        request_id: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        input_tokens: Option<u32>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        output_tokens: Option<u32>,
    },
    #[serde(rename_all = "camelCase")]
    Completed {
        request_id: String,
        finish_reason: FinishReason,
        total_ms: u64,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        time_to_first_token_ms: Option<u64>,
    },
    #[serde(rename_all = "camelCase")]
    Failed {
        request_id: String,
        error: BlueyError,
    },
}

/// Mirrors `AIResponse` (non-streaming convenience).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AiResponse {
    pub request_id: String,
    pub selection: ModelSelection,
    pub text: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub json: Option<serde_json::Value>,
    pub finish_reason: FinishReason,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input_tokens: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_tokens: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub time_to_first_token_ms: Option<u64>,
    pub total_ms: u64,
}

/// Mirrors `ConnectionTestResult`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectionTestResult {
    pub ok: bool,
    pub provider_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub latency_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<BlueyError>,
}

// ── Research ─────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum ResearchDepth {
    #[default]
    None,
    Search,
    SearchScrape,
    DeepAgent,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SearchSource {
    Exa,
    Firecrawl,
    Mock,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SearchResult {
    pub id: String,
    pub title: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub snippet: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub published_at: Option<String>,
    pub source: SearchSource,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScrapeResult {
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub markdown: String,
    pub source: SearchSource,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ResearchTool {
    ExaSearch,
    FirecrawlScrape,
    DocumentRead,
}

/// Mirrors `DeepResearchRequest`. `query`/`goal` must be public (no private context).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeepResearchRequest {
    pub job_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub query: String,
    pub goal: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_turns: Option<u32>,
    pub tools: Vec<ResearchTool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub allowed_document_ids: Option<Vec<String>>,
}

/// Mirrors `DeepResearchEvent` (tag = "type").
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DeepResearchEvent {
    #[serde(rename_all = "camelCase")]
    Started { job_id: String },
    #[serde(rename_all = "camelCase")]
    Progress { job_id: String, message: String },
    #[serde(rename_all = "camelCase")]
    ToolCall {
        job_id: String,
        tool: String,
        input: serde_json::Value,
    },
    #[serde(rename_all = "camelCase")]
    TextDelta { job_id: String, text: String },
    #[serde(rename_all = "camelCase")]
    Completed {
        job_id: String,
        report: String,
        citations: Vec<Citation>,
        total_ms: u64,
        turns: u32,
    },
    #[serde(rename_all = "camelCase")]
    Failed { job_id: String, error: BlueyError },
}
